Improve code snippets

hippotastic opened the pull request:
<!-- Thank you for opening a PR! We really appreciate you taking the time to help out 🙌 -->

#### What kind of changes does this PR include?
<!-- Delete any that don’t apply -->

- Changes to the docs site code

#### Description

- Automatically adds a file name tab (like in VS Code) to code snippets that contain a file name reference.
  - This unchanged code snippet from the existing docs site...
    ````md
    ```astro
    ---
    // Example: GreetingHeadline.astro
    // Usage: <GreetingHeadline greeting="Howdy" name="Partner" />
    const { greeting, name } = Astro.props
    ---
    <h2>{greeting}, {name}!</h2>
    ```
    ````
  - ...gets rendered like this:

    ![image](https://user-images.githubusercontent.com/6137925/180893175-6e6b997d-b9bd-4463-9758-ed9e6291afe0.png)

  - Tab titles can optionally be set/overridden in Markdown by using the meta information area after an opening code fence like this:
    ````md
    ```astro title="example.astro"
    <p>This is a demo.</p>
    ```
    ````

- Adds support for highlighting lines in code snippets by adding line ranges to the meta information area after an opening code fence.
  - This code snippet...
    ````md
    ```astro title="example.astro" {2,4-6}
    <p>This is a demo.</p>
    <p>This paragraph gets highlighted (line 2).</p>
    <p>This is a demo.</p>
    <p>This paragraph gets highlighted (line 4 is in range 4-6).</p>
    <p>This paragraph gets highlighted (line 5 is in range 4-6).</p>
    <p>This paragraph gets highlighted (line 6 is in range 4-6).</p>
    ```
    ````
   - ...gets rendered like this:
   
     ![image](https://user-images.githubusercontent.com/6137925/180894222-bec17a03-1383-4326-8403-c10c00b5896a.png)

- Automatically adds a Terminal style to shell code snippets:
  
  ![image](https://user-images.githubusercontent.com/6137925/180894376-ab830111-b1c2-423c-8f0e-643043cd5a6a.png)


<!--
Here’s what will happen next:

1. Our GitHub bots will run to check your changes.
   If they spot any broken links you will see some error messages on this PR.
   Don’t hesitate to ask any questions if you’re not sure what these mean!

2. In a few minutes, you’ll be able to see a preview of your changes on Netlify 🥳

3. One or more of our maintainers will take a look and may ask you to make changes.
   We try to be responsive, but don’t worry if this takes a day or two.
-->

sarah11918 commented:
These are looking snazzy, @hippotastic !  🔥 

Just noting here for the record about us finding the one-off case where a file name title was caught by mistake (Styling > load a static stylesheet via link tags) but otherwise, all other snippets passed! 🥳 

And, if you add some notes to `WRITING.md` for authors about how they should now write their code snippets, along where we document things like the custom asides, I'll do the "writing" part for it.

hippotastic commented:
Ok, so I've fixed the one edge case you mentioned. Here is the current summary of what has changed for authors (pending Chris' review):

- **You can add a title to any code snippet by specifying a `title="..."` metadata property on the opening code fence line. The title will be displayed as a tab.**
  - Example:
    ````md
    ```js title="example-file.js"
    // Some JS code
    ```
    ````

- **If no `title="..."` metadata property was specified, code snippets will look for a file name comment in the first 5 lines of the code, remove that line and use the file name as the title.**
  - In this example, the line `src/pages/MyAstroPage.astro` will be detected, removed and used as the code snippet title:
    ````md
    ```astro
    ---
    // src/pages/MyAstroPage.astro
    import MySidebar from '../components/MySidebar.jsx';
    ---
    <MySidebar>
      ...
    </MySidebar>
    ```
    ````

- **You can highlight lines in any code snippet by specifying the line numbers in a pair of curly braces on the opening code fence line.**
  - Line number syntax:
    - Single lines: `{1}` highlights the first line
    - Line ranges: `{3-5}` highlights lines 3, 4, and 5
    - Multiple lines/ranges separated by commas: `{1,4,7-10}` highlights lines 1, 4, and 7–10
  - Example:
    ````md
    ```astro title="example.astro" {2,4-6}
    <p>Regular line 1</p>
    <p>Highlighted line 2</p>
    <p>Regular line 3</p>
    <p>Highlighted line 4 (in range 4-6)</p>
    <p>Highlighted line 5 (in range 4-6)</p>
    <p>Highlighted line 6 (in range 4-6)</p>
    ```
    ````

- **Code snippets using a shell language (`bash`, `sh`, `shell`, `shellscript` or `zsh`) will automatically be styled like a Terminal window.**

delucis reviewed (commented):
Posting in progress comments. Main outstanding thing is checking there are no colour contrast issues in highlighted lines.

Other than you of course, Hippo — you are outstanding!

hippotastic commented on integrations/astro-code-snippets.ts:
```diff
@@ -0,0 +1,259 @@
+import { visit } from 'unist-util-visit';
+
+const CodeSnippetTagname = 'AutoImportedCodeSnippet';
+const LanguageGroups = {
+	code: ['astro', 'cjs', 'htm', 'html', 'js', 'jsx', 'mjs', 'svelte', 'ts', 'tsx', 'vue'],
+	data: ['env', 'json', 'yaml', 'yml'],
+	styles: ['css', 'less', 'sass', 'scss', 'styl', 'stylus'],
+	textContent: ['markdown', 'md', 'mdx'],
+};
```
I actually had a version that does not use hardcoded extensions before, but it caused various false positives, e.g. interpreting valid semver version numbers like `1.4.x` or `1.0.0-feature.rc` as file names. The language groups also help to avoid incorrectly considering our "edge case" comment to be a valid filename because `css` and `html` belong to different groups:

````md
```html
<head>
  <!-- Local: /public/styles/global.css -->
  <link rel="stylesheet" href="/styles/global.css" />
  <!-- External -->
  ...
</head>
```
````

I'm happy to take care of any updating that may be required for these lists. :)

hippotastic commented on integrations/astro-code-snippets.ts:
```diff
@@ -0,0 +1,259 @@
+	// Remove the matched string (if any) from meta
+	meta = (titleMatch?.[0] && meta.replace(titleMatch[0], '')) || meta;
+
+	// Parse numeric ranges like `{4-5,10}` into an array of highlighted lines (if any)
+	const highlightedLines = rangeParser(meta.match(/{(.*)}/)?.[1] || '');
```
The current syntax for line highlighting is based on the one used by `rehype-pretty-code`. I agree that we can explore other formats as well, e.g. if we want to allow "highlighted lines", "highlighted words", "focused lines" (others could be collapsed in long snippets) etc. - I'm very open to any improvements in the future!

Yan-Thomas commented on public/theme.css:
```diff
@@ -116,11 +116,14 @@
 	--theme-selection-color: hsla(var(--color-purple), 1);
 	--theme-selection-bg: hsla(var(--color-purple), var(--theme-accent-opacity));
+	--theme-code-selection-bg: hsla(var(--color-purple), 0.4);
+	--theme-code-highlight-bg: hsla(var(--color-purple), 0.3);
```
```suggestion
	--theme-code-highlight-bg: hsla(var(--color-purple), 0.2);
```

hippotastic replied:
Thank you for testing this! I brought the border back as you suggested and fixed background colors until Chrome's built-in color contrast checker was happy. I had to go even further down than 0.2 to make Chrome happy.

hippotastic commented on src/components/CodeSnippet.astro:
```diff
@@ -0,0 +1,145 @@
+	});
+}
+---
+<style lang="scss" is:global>
+	.code-snippet {
```
I tried it without `is:global`, adding `:global()` around various selectors, but couldn't get it to work. Maybe we still have an Astro issue here, yes. Looking forward to improving this in the future if we find a way!

hippotastic commented on src/components/CodeSnippet.astro:
```diff
@@ -0,0 +1,145 @@
+		titleHtml && <span class="title" set:html={titleHtml} />
+	}</div>
+	<Fragment set:html={codeSnippetHtml} />
+</div>
```
Ha! I love it!

Yan-Thomas reviewed (commented):
I did some accessibility auditing in this PR, the only failing AA color contrast was between the highlight color and comments color, tweaking the light theme highlight a bit fixed it.

In your first screenshot there was a border in the highlighted lines, was there a reason to remove it? If not, I think it would be nice to bring it back, having it makes the solution accessible for color blind users and makes us need to worry less about the contrast between the highlighted and normal background colors, since we have an extra visual cue.
